src/MainWindow.vala

```diff
@@ -877,30 +793,50 @@ namespace Terminal {
             }
         }
 
+        private bool close_immediately = false;
         //TODO Make TerminalWidget.confirm_kill_fg_process asynchronous and terminate all in callback
         public bool on_delete_event () {
+            if (close_immediately) {
+                return Gdk.EVENT_PROPAGATE;
+            }
             //Avoid saved terminals being overwritten when tabs destroyed.
             notebook.tab_view.page_detached.disconnect (on_tab_removed);
             save_opened_terminals (true, true);
-            var tabs_to_terminate = new GLib.List <TerminalWidget> ();
-
             for (int i = 0; i < notebook.n_pages; i++) {
                 var term = get_term_widget (notebook.tab_view.get_nth_page (i));
-                if (term.confirm_kill_fg_process (
-                    _("Are you sure you want to quit Terminal?"),
-                    _("Quit Terminal"))
-                ) {
-                    tabs_to_terminate.append (term);
-                } else {
-                    return true;
+                if (term.has_foreground_process ()) {
+                    term.confirm_kill_fg_process (
+                        _("Are you sure you want to close all foreground processes before closing the window?"),
+                        _("Close window"),
+                        (() => {
+                            terminate_all ();
+                            close_immediately = true;
+                            this.close ();
+                        })
+                    );
+
+                    return Gdk.EVENT_STOP;
                 }
             }
 
-            foreach (var t in tabs_to_terminate) {
-                t.term_ps ();
+            terminate_all ();
```

Comment:
Won't this just run immediately without waiting for confirmation? This seems like it should be in an early return instead

Reply:
This is intended to run *after* the user has confirmed that all processes should be terminated. I did not think a separate confirmation for each tab desirable (although I guess with extra code it could be accomodated).  I'll check that the behaviour is as intended.